tikv/raftstore/fsm.go

```diff
@@ -0,0 +1,556 @@
+	mu      *sync.RWMutex
+	normals map[uint64]*mailbox
+	caches  map[uint64]*mailbox
+
+	controlBox *mailbox
+
+	controlScheduler fsmScheduler
+	normalScheduler  fsmScheduler
+}
+
+func newRouter(controlBox *mailbox, controlScheduler, normalScheduler fsmScheduler) *router {
+	return &router{
+		mu:               new(sync.RWMutex),
+		normals:          make(map[uint64]*mailbox),
+		caches:           make(map[uint64]*mailbox),
+		controlBox:       controlBox,
+		controlScheduler: controlScheduler,
+		normalScheduler:  normalScheduler,
+	}
+}
+
+func (r *router) register(addr uint64, mailbox *mailbox) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	old := r.normals[addr]
+	r.normals[addr] = mailbox
+	if old != nil {
+		old.close()
+	}
+}
+
+type addrWithMailbox struct {
+	addr    uint64
+	mailbox *mailbox
+}
+
+func (r *router) registerAll(mailboxes []addrWithMailbox) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	for _, mb := range mailboxes {
+		old := r.normals[mb.addr]
+		r.normals[mb.addr] = mb.mailbox
+		if old != nil {
+			old.close()
+		}
+	}
+}
+
+func (r *router) mailbox(addr uint64) *mailbox {
+	caches := r.caches
+	mb, ok := caches[addr]
+	if ok {
+		if atomic.LoadInt64(&mb.state.status) == notifyStateDrop {
+			delete(caches, addr)
+			return nil
+		}
+		return mb
+	}
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	mb, ok = r.normals[addr]
+	if !ok {
+		return nil
+	}
+	caches[addr] = mb
+	return mb
+}
+
+func (r *router) send(addr uint64, msg *Msg) error {
+	mb := r.mailbox(addr)
+	if mb == nil {
+		return errors.Errorf("mailbox is nil for addr %d", addr)
+	}
+	mb.send(msg, r.normalScheduler)
+	return nil
+}
+
+func (r *router) sendControl(msg *Msg) {
+	r.controlBox.send(msg, r.controlScheduler)
+}
+
+func (r *router) broadcastShutdown() {
+	log.Info("broadcasting shutdown")
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	for addr, mb := range r.normals {
+		log.Debugf("region %d shutdown mailbox", addr)
+		mb.close()
+	}
+	r.controlScheduler.shutdown()
+	r.normalScheduler.shutdown()
+}
+
+func (r *router) close(addr uint64) {
+	log.Infof("region %d shutdown mailbox", addr)
+	delete(r.caches, addr)
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	mb, ok := r.normals[addr]
+	if ok {
+		mb.close()
+		delete(r.normals, addr)
+	}
+}
+
+func (r *router) clone() *router {
+	cloned := *r
+	cloned.caches = make(map[uint64]*mailbox)
+	return &cloned
+}
+
+type batch struct {
+	normals []fsm
+	control fsm
+}
+
+func newBatchWithCapacity(cap int) *batch {
+	return &batch{
+		normals: make([]fsm, 0, cap),
+	}
+}
+
+func (b *batch) push(msg *Msg) {
+	switch msg.Type {
+	case MsgTypeFsmNormal:
+		b.normals = append(b.normals, msg.Fsm)
+	case MsgTypeFsmControl:
+		y.Assert(b.control == nil)
+		b.control = msg.Fsm
+	default:
+		panic(fmt.Sprintf("unexpected msg type %d", msg.Type))
+	}
+}
+
+func (b *batch) len() int {
+	l := len(b.normals)
+	if b.control != nil {
+		l++
+	}
+	return l
+}
+
+func (b *batch) isEmpty() bool {
+	return b.len() == 0
+}
+
+func (b *batch) clear() {
```

Comment:
Should we need to add a shrink function to reclaim the memory?

Reply:
No the operation `b.normals = b.normals[:0]` reuses the memory.